### PR TITLE
Add Spring Data Jest to improve Elasticsearch

### PR DESCRIPTION
**Pull request by mraible**
- Please make sure the below checklist is followed for Pull Requests.

- [x] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
- [x] Tests are added where necessary
- [x] Documentation added in https://github.com/jhipster/jhipster.github.io/pull/613
- [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

Implements changes in #7810.

### CODE REVIEW

**Comment by mraible** [comment]
The Travis build for this PR keeps failing with errors like:

```
  JHipster generator
    Default configuration with
      AngularX
keytool error: java.io.FileNotFoundException: src/main/resources/config/tls/keystore.p12 (No such file or directory)
        1) "before each" hook
      React
keytool error: java.io.FileNotFoundException: src/main/resources/config/tls/keystore.p12 (No such file or directory)
        2) "before each" hook
      using npm flag
keytool error: java.io.FileNotFoundException: src/main/resources/config/tls/keystore.p12 (No such file or directory)
        3) "before each" hook
      Gradle
```

@jdubois Is this because of a change you recently made on the master branch?

File: generators/cleanup.js [hunk 1]
```diff
@@ -156,4 +156,7 @@ function cleanupOldServerFiles(generator, javaDir, testDir, mainResourceDir, tes
         generator.removeFile(`${testDir}web/rest/ProfileInfoResourceIntTest.java`);
         generator.removeFile('gradle/mapstruct.gradle');
     }
+    if (generator.isJhipsterVersionLessThan('5.2.1')) {
```
Comment (mraible) [inline generators/cleanup.js]: @vishal423 Is this the correct version to use? `ElasticsearchConfiguration.java` is removed in this PR and no longer needed in 5.2.2+.

**Comment by jdubois** [comment]
Sorry @mraible , it looks like I broke the build, I'm having a look right now

**Comment by jdubois** [comment]
Master should be fixed now, sorry :-(

**Comment by mraible** [comment]
FWIW, I tried to remove Log4J as a dependency using [this guide](https://www.elastic.co/guide/en/elasticsearch/client/java-api/current/_using_another_logger.html), but failed.

**Comment by jdubois** [comment]
@mraible oh I remember trying (and failing!) this also!

File: generators/cleanup.js (hunk 1 above)
Comment (vishal423) [reply generators/cleanup.js]: looks good

**Comment by jdubois** [comment]
This means people won't be able to use Spring Data Elasticsearch at all. As it's the official implementation, I'm not sure this is a good idea. I understand it won't work on cloud providers, and I agree Spring Data Jest looks like a better solution, but still that's an important change.
Besides, shouldn't you also remove Spring Data Elasticsearch from the Maven/Gradle dependencies?

File: generators/cleanup.js (hunk 1 above)
Comment (jdubois) [reply generators/cleanup.js]: yes this API is really confusing!! I'm not totally sure, shouldn't that be 5.2.2?

File: generators/server/templates/src/main/resources/config/application-dev.yml.ejs [hunk 2]
```diff
@@ -193,12 +193,9 @@ spring:
     <%_ } _%>
     <%_ if (searchEngine === 'elasticsearch') { _%>
         elasticsearch:
-            cluster-name:
-            cluster-nodes: localhost:9300
             properties:
                 path:
-                    logs: <%= BUILD_DIR %>elasticsearch/log
-                    data: <%= BUILD_DIR %>elasticsearch/data
+                    home: <%= BUILD_DIR %>elasticsearch
```
Comment (jdubois) [inline generators/server/templates/src/main/resources/config/application-dev.yml.ejs]: as this is a Spring Data Elasticsearch property, is this being used by Jest also?

Comment (mraible) [reply generators/server/templates/src/main/resources/config/application-dev.yml.ejs]: Yes. Spring Data Jest doesn’t reimplement anything in Spring Data Elasticsearch. It uses this property to create the embedded ES instance. If you don’t have spring.data.jest.uri defined (which we don’t in the dev profile), this property is required.

**Comment by mraible** [comment]
Nothing changes from a code perspective. You can still use Spring Data Elasticsearch. Dependencies change and the port changes from 9300 (binary API) to 9200 (REST API), but that’s about it. I think it’s a good idea because: 1. It works and solves Elasticsearch on Heroku. 2. It allows for embedded Elasticsearch which is a much better development experience. 3. If people want to switch back, they only need to remove the dependency and change a couple properties.

File: generators/cleanup.js (hunk 1 above)
Comment (mraible) [reply generators/cleanup.js]: That’s sounds more logical. I’ll fix it.

**Comment by jdubois** [comment]
Thanks @mraible - you're right, let's do this! So much better than what is provided by default!

File: generators/cleanup.js (hunk 1 above)
Comment (vishal423) [reply generators/cleanup.js]: yeah, looks like I mis-interpreted the api. I am bit lost in jhipster versioning :confused:

**Comment by jdubois** [comment]
Matt, doesn't this need a new release of the JHipster dependencies?

**Comment by DanielFran** [comment]
@jdubois For now it is not needed since @mraible added the version in the commits.

So for now it is not dependent of https://github.com/jhipster/jhipster/pull/87

**Comment by jdubois** [comment]
Oh thanks @DanielFran - let me release a new version of JHipster dependencies, so those could be removed.

**Comment by DanielFran** [comment]
@jdubois Do you want I split my PR in jhispter in 2, 1 that we could already aplly in this new version and another one with specific related with spring 2.1.x?

**Comment by jdubois** [comment]
Thanks @DanielFran but it's already released :-)
However I have a couple of issues, see https://github.com/jhipster/generator-jhipster/issues/7810#issuecomment-415354308

**Comment by cbornet** [comment]
Does it mean we can now use an embedded ES in the tests like we used to instead of the current mock-based implementation ?

**Comment by jdubois** [comment]
@cbornet yes I guess we can do it again now.

**Pull request merged**